src/libraries/System.Net.Security/src/System/Net/Security/NegotiateStream.cs

```diff
@@ -720,97 +740,46 @@ private async Task AuthenticateAsync<TIOAdapter>(CancellationToken cancellationT
-                    exception = new AuthenticationException(SR.net_auth_bad_client_creds_or_target_mismatch);
-                    int statusCode = ERROR_TRUST_FAILURE;
-                    message = new byte[sizeof(long)];
-
-                    for (int i = message.Length - 1; i >= 0; --i)
-                    {
-                        message[i] = (byte)(statusCode & 0xFF);
-                        statusCode = (int)((uint)statusCode >> 8);
-                    }
-
-                    await SendAuthResetSignalAndThrowAsync<TIOAdapter>(message, exception, cancellationToken).ConfigureAwait(false);
-                    Debug.Fail("Unreachable");
-                }
-
-                if (PrivateImpersonationLevel < _expectedImpersonationLevel)
-                {
-                    exception = new AuthenticationException(SR.Format(SR.net_auth_context_expectation, _expectedImpersonationLevel.ToString(), PrivateImpersonationLevel.ToString()));
-                    int statusCode = ERROR_TRUST_FAILURE;
-                    message = new byte[sizeof(long)];
-
-                    for (int i = message.Length - 1; i >= 0; --i)
-                    {
-                        message[i] = (byte)(statusCode & 0xFF);
-                        statusCode = (int)((uint)statusCode >> 8);
-                    }
-
-                    await SendAuthResetSignalAndThrowAsync<TIOAdapter>(message, exception, cancellationToken).ConfigureAwait(false);
-                    Debug.Fail("Unreachable");
-                }
-
-                ProtectionLevel result = _context.IsConfidentialityFlag ? ProtectionLevel.EncryptAndSign : _context.IsIntegrityFlag ? ProtectionLevel.Sign : ProtectionLevel.None;
+                    NegotiateAuthenticationStatusCode.BadBinding =>
+                        new AuthenticationException(SR.net_auth_bad_client_creds_or_target_mismatch),
+                    NegotiateAuthenticationStatusCode.TargetUnknown =>
+                        new AuthenticationException(SR.net_auth_bad_client_creds_or_target_mismatch),
+                    NegotiateAuthenticationStatusCode.ImpersonationValidationFailed =>
+                        new AuthenticationException(SR.Format(SR.net_auth_context_expectation, _expectedImpersonationLevel.ToString(), PrivateImpersonationLevel.ToString())),
+                    _ => // NegotiateAuthenticationStatusCode.SecurityQosFailed
+                        new AuthenticationException(SR.Format(SR.net_auth_context_expectation, _context.ProtectionLevel.ToString(), _expectedProtectionLevel.ToString())),
+                };
 
-                if (result < _expectedProtectionLevel)
-                {
-                    exception = new AuthenticationException(SR.Format(SR.net_auth_context_expectation, result.ToString(), _expectedProtectionLevel.ToString()));
-                    int statusCode = ERROR_TRUST_FAILURE;
-                    message = new byte[sizeof(long)];
+                message = new byte[sizeof(long)];
```

Comment:
can we make it static constant?

Reply:
I suppose we can but I am not sure it's worth it. It's an error condition that can happen only once per connection. Is it really worth optimizing an error code path?

Reply:
while the impact is small it also seem super easy...

Reply:
It was not done previously. The code basically stayed the same. I felt it was hard enough to read as-is.